docs/pwo2xsf.1
.TH pwo2xsf 1 "August 7, 2019" "XCrySDen" "XCrySDen"

.SH NAME
pwo2xsf \- convert the pw.x output file to XSF format

.SH SYNOPSIS
.B pwo2xsf
.RI [ special-option ] 
\fIcoordinates-option\fR
[\fIfile\fR]

.SH DESCRIPTION
\fBpwo2xsf\fR extracts atomic structures from \fBpw.x\fR output
file and prints the XSF formated structure data to standard output.
It is a utility program of \fBxcrysden\fR package
(\fIhttp://www.xcrysden.org/\fR) and XSF is the native \fBxcrysden
structure format\fR (\fIhttp://www.xcrysden.org/doc/XSF.html\fR).

If \fIfile\fR is not specified then standard input is read.

.SH SPECIAL OPTION

.TP 4
\fB\-r \fIdim\fR
Reduce periodic dimension to \fIdim\fR, i.e., \fIdim\fR = 2, 1, or 0.

.SH COORDINATES OPTIONS

These options specify which cordinates to extract. The possibilities
are:

.TP 4
\fB\-ic \fIfile\fR 
.PD 0 
.TP 
\fB\-\-inicoor \fIfile\fR 
Extract the initial, that is, input ionic coordinates.

.TP 4
\fB\-lc \fIfile\fR 
.PD 0 
.TP 
\fB\-\-latestcoor \fIfile\fR 
Extract the latest printed ionic coordinates from pw.x output file.

.TP 4
\fB\-oc \fIfile\fR 
.PD 0 
.TP 
\fB\-\-optcoor \fIfile\fR
Similar to "\-\-latestcoor" but extracts just the optimized coordinates
if present or else nothing.

.TP 4
\fB\-a \fIfile\fR 
.PD 0 
.TP 
\fB\-\-animxsf \fIfile\fR
Extract the coordinates of all ionic steps as Animated-XSF.

.TP 4
\fB\-nc file\fR 
.PD 0 
.TP 
\fB\-\-nebcoor PW.out\fR
Extract the coordinates and forces from the \fBPW.out\fR file of NEB
(path) calculation.


.SH "COPYRIGHT"

Copyright (C) 1996--2019 Anton Kokalj


This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.  You should have received a copy of
the GNU General Public License along with this program; if not, write
to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.



NEWS
Version: 1.6:
-------------

* "Reload" button added to the main window (at the bottom toolbar)
  that reloads the structure from file; useful for structure building

* lattice-vector reduction (Wigner-Seitz cell and Brilluouin-zone were
  incorrect for non-reduced lattice-vectors)

* compatibility with Quantum ESPRESSO (QE) 6.x, in particular:
  - display of structures from neb.x input file
  - aware of the new ibrav lattice definitions of QE (e.g. negative
    ibrav and ibrav=91);
  - display of conventional cell-mode for centered lattice display
    from pw.x output    
  - compatible with "ATOMIC_POSITIONS (crystal_sg)" specs, but it
    needs pw.x on $PATH

* compatibility with CRYSTAL17

* ported to Tcl/Tk-8.6 (thanks to Julien Lefevre) and Togl-2.0

* because the API of Togl2.0 changed, the syntax of *.xcrysden scripts
  changed as well (for the Togl part); old *.xcrysden scripts can be
  loaded with --os or --oldscript command line option

* Togl options can be specified in ~/.xcrysden/custom-definitions file
  (e.g. if xcrysden crashes with "X Error of failed request: ...", try
  setting "set toglOpt(accum) false"; but beware that this disables
  anti-aliasing)

* if "Print To File" produces blank images, add "set printSetup(dumpWindow) 1"
  to ~/.xcrysden/custom-definitions file

* correct illumination of isosurfaces even if the DATAGRID box is
  left-oriented (cube files in Gaussian use left-oriented boxes);

* new "File-->Save Current State" menu entry. It saves only a display
  "state" of the currently displyed structure; this is useful for
  scripting (see below item)

* new "--print" command line option for batch printing. This is useful
  within the shell scripts to achieve "automatic" printing. Example:
  xcrysden --xsf file.xsf --script state.xcrysden --print file.png

* bug-fix: Wien2K examples (File-->XCrySDen Examples ...-->WIEN2K Struct Files)
  did not work if xcrysden was installed to read-only path


Version: 1.5:
-------------

* display of structure from Orca output

* new Movie-Maker: xcrysden allows for real-time capture of display. Three
  modes are supported: "real-time capture", "every redisplay", and
  "manual".

* several movie encoders are supported, in particular

  - for Animated-GIF: convert (from ImageMagick), gifsicle, and
    whirlgif; convert is used by default. 

  - for AVI/MPEG: mencoder (from mplayer) and ppmtompeg. Mencoder is
    used by default.

* better algorithm for isosurface shading (nicer surfaces)

* no more external package specifications in
  ~/.xcrysden/custom-definitions, due to automatic package finding
  mechanism. The only exception is the CRYSTAL package
  (http://www.crystal.unito.it/), which still needs to be explicitly
  defined.

* customization of Fermi surface viewer is possible in ~/.xcrysden/custom-definitions

* new command-line option --custom for reading customization from specified file

* compatibility with Quantum-Espresso 5.x, WIEN2k_08+, and CRYSTAL14

* XCRYSDEN_TOPDIR and XCRYSDEN_SCRATCH environmental variables are not
  required anymore. The use of the former is deprecated and the latter
  is set to /tmp if unspecified.

* the use of "xcConfigure" is deprecated

* "make install" now installs the package system-wide (see README file
  for more details).

* ~/.xcrysden/Xcrysden_resources is used instead of
  ~/.xcrysden/Xcrysden_defaults. This is due to a new font handling
  mechanism in The Tcl/Tk 8.5, which makes the fonts as specified by
  old ~/.xcrysden/Xcrysden_defaults to look pretty ugly.

* improved measurement procedure: the selection of atoms is more
  precise (i.e. easier) and the coordinates of mouse-click selected
  atoms can be displayed in Angstrom, Bohr, crystal-primitive,
  crystal-conventional, and alat units. Measured distances are
  reported also in bohr units, and the printout format of measured
  distances and angles is configurable

* adding mouse-wheel support for display-zooming (I finally got a
  mouse with a wheel).

* a "stop-playing" button added to animations control window 

* crystal borders are displayed by default for periodic structures

* checking for stale files in $XCRYSDEN_SCRATCH (prompts the user for their deletion)

* only one flavor of source package (external library sources are
  download and compiled on request)

* many bug fixes 

* for further information, see ChangeLog


Version: 1.4:
-------------

* automatic labeling of k-points for k-path selection
  (thanks to Peter Blaha)

* Stereo display mode (thanks to Gerardo Ballabio)

* Anaglyph display mode, i.e., fake stereo, requires red-blue glasses
  (thanks to Eric VERFAILLIE)

* a few Bug fixes in visualization of Fermi surfaces (thanks to Pablo
  de la Mora)

* improved EPS printing (also PDF printing) thanks to new
  gl2ps-1.2.4. The vectorial EPS printing of Lighting-On mode is
  finally good enough.

* comments added to the XSF format; comment-lines start with the "#"
  character and can be located only in between the sections, but
  within a given section.


Version: Beta-1.2
-----------------

* improved visualization of Fermi surfaces (a few Bug fixes; thanks to Pablo de la Mora)

* visualization of vector-fields

* dummy "X" atom does not have bonds anymore

* multi-band display of Fermi surfaces in one widget

* added support for CRYSTAL03

* more configurable options for forces (color, thickness)

* Coordinate system color is configurable


Version: Beta-1.0
-----------------

* XCRYSDEN covered by GNU GENERAL PUBLIC LICENSE

* XCRYSDEN ported to MAC OSX (requires Xserver)

* XCRYSDEN ported to MS-Windows (under CYGWIN). Does not require
  X-server, but require CYGWIN (http://www.cygwin.com/).

* Visualization of H-bonds (lighting-on mode only).

* Update of the PWscf IO filters (compatible with PWscf v.2.0).

* XCRYSDEN interface was mad compatible with CRYSTAL03

* Simple bond editing: the covalent-radius item
  (Modify-->Atomic Radius menu) has been split into

scripts/pwo2xsf.sh
#!/bin/sh
#############################################################################
# Author:                                                                   #
# ------                                                                    #
#  Anton Kokalj                                  Email: Tone.Kokalj@ijs.si  #
#  Department of Physical and Organic Chemistry  Phone: x 386 1 477 3523    #
#  Jozef Stefan Institute                          Fax: x 386 1 477 3811    #
#  Jamova 39, SI-1000 Ljubljana                                             #
#  SLOVENIA                                                                 #
#                                                                           #
# Source: $XCRYSDEN_TOPDIR/scripts/pwo2xsf.sh
# ------                                                                    #
# Copyright (c) 1996-2003 by Anton Kokalj                                   #
#############################################################################

# set locales to C
LANG=C 
LC_ALL=C
export LANG LC_ALL

#
# Purpose: PWscf(v2.0 or latter)-output--to--XSF conversion
# Usage:   pwo2xsf [options] pw-output-file
#
# Written by Tone Kokalj on Mon Feb  9 12:48:10 CET 2004
#            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

cat > pwo2xsfUsage.$$ <<EOF

 Usage: pwo2xsf.sh [-r dim] options [pw-output-file]

 If file is not specified then stadard input is read.

  -r dim 
     Reduce periodic dimension to dim, where dim = 2, 1, or 0.
 
 OPTIONS are:

  --inicoor|-ic [pw-output-file]
     Extract the initial (i.e. input) ionic coordinates.

  --latestcoor|-lc [pw-output-file]
     Extract the latest printed ionic coordinates from pw.x output file.

  --optcoor|-oc [pw-output-file]
     Similar to "--latestcoor" but extracts just the optimized coordinates
     if present or else nothing.

  --animxsf|-a [pw-output-file1] ...
     Extract the coordinates of all ionic steps as Animated-XSF.

  --nebcoor|-nc [PW.out]
     Extract the coordinates and forces from the \fBPW.out\fR file of NEB
     (path) calculation.

EOF


# ------------------------------------------------------------------------
# Utility functions
# ------------------------------------------------------------------------
CleanTmpFiles() {
    if test -f pwo2xsfUsage.$$ ; then rm -f pwo2xsfUsage.$$; fi
    if test -f xsf.$$ ; then rm -f xsf.$$; fi
    if test -f pw.$$ ; then rm -f pw.$$; fi
}
pwoExit() {
    # Usage: $0 exit_status
    CleanTmpFiles
    exit $1
}
pwoUsage() {
    if test $1 ; then
	echo "
Usage: $2
"
	pwoExit 1
    fi
}
pwoGetVersion() {
    grep 'Program PWSCF' $1 | tail -1 | awk '
      {  ver=$3; n=split(ver,v,"."); 
         ii=0; for(i=1; i<=n; i++) if ( v[i] ~ /[0-9]+/ ) vv[ii++]=v[i];  
         if (ii>2) printf "%d.%d%d\n", vv[0],vv[1],vv[2];
         else printf "%d.%d\n", vv[0], vv[1];
      }'
}
pwoCheckPWSCFVersion() {
    #
    # Usage: $0 option file
    #
    # Purpose: if PWSCF version < 1.3 execute the old pwo2xsf_old.sh
    #          script and exit
    version=`pwoGetVersion $input`
    result=`echo "$version < 1.3"|bc -l`
    if test x"$result" = x1 ; then
	if test -f $scriptdir/pwo2xsf_old.sh ; then
    	   # execute pwo2xsf_old.sh
	    $scriptdir/pwo2xsf_old.sh $1 $2
	    pwoExit $?
	else
	    echo "ERROR: PWscf output generated by version < 1.3 !!!"
	    pwoExit 1
	fi
    fi
}


# ------------------------------------------------------------------------
# Function: pwoGetIbrav
# Extract:  ibrav from pw.x output file
# ------------------------------------------------------------------------
pwoGetIbrav() {
    echo `egrep "bravais-lattice index  +=" $1 | tail -1 | awk '{print $NF}'`
}

pwoAnimSteps() {
    cp xsf.$$ xsf2.$$
    nsteps=`grep PRIMCOORD xsf.$$ | wc | awk '{print $1}'`    
    echo "ANIMSTEPS $nsteps" > xsf.$$
    cat xsf2.$$ >> xsf.$$
    rm -f xsf2.$$
}

# ------------------------------------------------------------------------
# Function: pwoConvVec
# Purpose: tries to assign the CONVVEC based on PRIMVEC and ibrav and
#          prints to standard output the modified XSF file that
#          contains the CONVVECs
#------------------------------------------------------------------------
pwoConvVec() {
    ibrav=`pwoGetIbrav $input`

    if test "x$ibrav" = x; then
        # ibrav not found; do nothing
        cat xsf.$$
        
    elif test "x$dim" = x3; then
        # ibrav was found and structure was not reduced; try to assign CONVVEC
        $PWO_XSF2XSF $ibrav < xsf.$$
        
    else
        # dimensionality of structure has bin reduced; do nothing
        cat xsf.$$
    fi
}


# ------------------------------------------------------------------------
# Function: pwoOptCoor
# Extract:  OPTIMIZED or LATEST coordinates
# Perform:  read PW-output file and print the XSF file according to
#           specified flags
# ------------------------------------------------------------------------
pwoOptCoor() {
    #set -x
    pwoUsage "$# -lt 1" \
	"$0 --latestcoor|-lc [pw-output-file]   or   $0 --optcoor|-oc [pw-output-file]"
    
    option=$1
    case $1 in
	--latestcoor|-lc) type=LATEST;    shift;;
	--optcoor|-oc)    type=OPTIMIZED; shift;;
    esac
    
    if test $# -eq 0 ; then
	input=pw.$$
	cat - >> $input
    else
	input=$1
    fi
    
    pwoCheckPWSCFVersion $option $input

    if test $type = "OPTIMIZED" ; then
	# Check for the presence of "Begin final coordinates"
	# and/or:
	# Check also for the PWSCF-v.1.3.0 which uses the
	# "Final estimate of positions" record
	if test \( "`grep \"Begin final coordinates\" $input`" = "" \) \
	    -a \( "`grep 'Final estimate of positions' $input`" = "" \) \
	    -a \( "`grep '     bfgs converged in' $input`" = "" \); then
	    echo "ERROR: OPTIMIZED coordinates does not exists"
	    pwoExit 1
	fi
    fi

    cat "$input" | awk -v t=$type -v dimen=$dim -f $scriptdir/pwo2xsf_opt.awk > xsf.$$

    pwoConvVec
}


# ------------------------------------------------------------------------
# Function: pwoAnimCoor
# Extract:  INITIAL or ALL coordinates
# Perform:  read PW-output file and print the XSF file according to
#           specified flags
# ------------------------------------------------------------------------

pwoAnimCoor() {
    #set -x
    pwoUsage "$# -lt 1" "$0 --animcoor|-ac|--animxsf|-a [pw-output-file1] or   $0 --inicoor|-ic [pw-output-file]"
    
    option=$1
    only_init=0
    case $1 in
	--inicoor|-ic) only_init=1; shift;;
	--animcoor|-ac|--animxsf|-a) only_init=0; shift;;
    esac

    if test $# -eq 0 ; then
	input=pw.$$
	cat - >> $input
    else
	input=$1
    fi

    pwoCheckPWSCFVersion $option $input

    ncoor=`awk '$1 ~ /ATOMIC_POSITIONS/ { npos++; } 
               END {   # add another step for initial coordinates    
                       print npos+1; }' $input`; 

    nvec=`awk '$1 ~ /CELL_PARAMETERS/ { nvec++; } END { print nvec; }' $input`
    
    cat "$input" | awk \
	-v ncoor=$ncoor \
	-v nvec=$nvec \
	-v dimen=$dim \
	-v onlyinit=$only_init -f $scriptdir/pwo2xsf_anim.awk > xsf.$$

    if test $only_init -eq 0 ; then
        # Assign the number of ANIMSTEPS here. The reason is that the
        # output file (queue runs) is the result of several job runs,
        # then some of them might be terminated on the "wrong" place,
        # and the initial ANIMSTEPS might be wrong. The most secure
        # way is to extract the sequential digit from the last
        # "PRIMCOORD id" record.
        pwoAnimSteps
    fi

    pwoConvVec
}

# ------------------------------------------------------------------------
# Function: pwoNebCoor
# Extract:  all coordinates and forces found in PW.out file
# ------------------------------------------------------------------------

pwoNebCoor() {
    #set -x
    pwoUsage "$# -lt 1" "$0 --nebcoor|-nc [PW.out]"
    
    option=$1; shift

    if test $# -eq 0 ; then
	input=pw.$$
	cat - >> $input
    else
	input=$1
    fi

    cat "$input" | awk \
	-v dimen=$dim \
	-f $scriptdir/pwo2xsf_neb.awk > xsf.$$

    # Assign the number of ANIMSTEPS here. The reason is that the
    # output file (queue runs) is the result of several job runs, then
    # some of them might be terminated on the "wrong" place, and the
    # initial ANIMSTEPS might be wrong. The most secure way is to extract the 
    # sequential digit from the last "PRIMCOORD id" record.

    pwoAnimSteps

    # conventional-vectors 
    pwoConvVec
}


#######################################################################
####                              MAIN                              ###
#######################################################################

if test "x`which readlink`" = "x"; then
    # no readlink cmd; make a function-substitute
    readlink_f() {
        file=`ls -l "$1" | awk '{print $NF}'`
        while test -h "$file"; do
            file=`ls -l "$file" | awk '{print $NF}'`
        done
        echo $file
    }
else
    readlink_f() {
        readlink -f "$1"
    }
fi


pathname() {
    file=`which "$1"`    
    if test $? -gt 0; then
        file=`type "$1"`
	if test $? -gt 0; then
	    # give-up
	    file="$1"
        else
            file=`echo $file | awk 'BEGIN {FS="is "} {print $NF}'`
        fi
    fi
    echo $file
}


pathdir() {
    file=`pathname "$1"`
    
    while test -h "$file"; do
	file=`readlink_f "$file"`
    done

    dir=`dirname "$file"`
    ( cd $dir; pwd )
}

if test -z $XCRYSDEN_TOPDIR; then
    # XCRYSDEN_TOPDIR does not exists, guess it from the process
    scriptdir=`pathdir $0`
    export XCRYSDEN_TOPDIR=`(cd $scriptdir/..; pwd)`
else
    scriptdir=$XCRYSDEN_TOPDIR/scripts
fi


AWK=`which awk`
if test "$AWK" = ""; then 
    echo "ERROR: awk program not found"
    pwoExit 1
fi

PWO_XSF2XSF=pwo_xsf2xsf
if test -f $XCRYSDEN_TOPDIR/bin/pwo_xsf2xsf ; then
    PWO_XSF2XSF=$XCRYSDEN_TOPDIR/bin/pwo_xsf2xsf
elif test -f $XCRYSDEN_LIB_BINDIR/pwo_xsf2xsf ; then
    PWO_XSF2XSF=$XCRYSDEN_LIB_BINDIR/pwo_xsf2xsf
fi

if [ $# -eq 0 ]; then
    cat pwo2xsfUsage.$$
    pwoExit 1
fi

dim=3
if test $1 = "-r"; then
    dim=$2;
    shift 2
fi

case $1 in
    --inicoor|-ic)               pwoAnimCoor $@;;
    --latestcoor|-lc)            pwoOptCoor $@;;
    --optcoor|-oc)               pwoOptCoor $@;;
    --animcoor|-ac|--animxsf|-a) pwoAnimCoor $@;;    
    --nebcoor|-nc)               pwoNebCoor $@;;
    --help|-h)                   cat pwo2xsfUsage.$$;;
    *)                           cat pwo2xsfUsage.$$; pwoExit 1;;
esac

pwoExit 0
